In [7]:
import pandas as pd 
import collections

In [16]:
df = pd.read_csv('invariantApis.csv')
df


Name     Return Type  \
0            onLivePusherBGMProgress             NaN   
1                   onAddToFavorites             NaN   
2           showNavigationBarLoading  <Primordial,V>   
3                         onMusicEnd             NaN   
4                           saveFile  <Primordial,V>   
...                              ...             ...   
1022         openChannelsOrderCenter  <Primordial,V>   
1023         openWCPayLQTDepositPlan  <Primordial,V>   
1024    startFacialRecognitionVerify  <Primordial,V>   
1025                   shareTimeline  <Primordial,V>   
1026  enableLocationUpdateBackground  <Primordial,V>   

                                             Class Name Method Name  \
0                                 <Application,Lgt0/s0>         NaN   
1     <Application,Lcom/tencent/mm/plugin/appbrand/m...         NaN   
2                                 <Application,Lzt0/s1>           u   
3     <Application,Lcom/tencent/mm/plugin/appbrand/j...         NaN   
4     <Application,Lcom/tencent/mm/plugin/appbrand/j...           u   
...                                                 ...         ...   
1022  <Application,Lcom/tencent/mm/plugin/appbrand/j...           u   
1023  <Application,Lcom/tencent/mm/plugin/appbrand/j...           u   
1024  <Application,Lcom/tencent/mm/plugin/appbrand/j...           u   
1025  <Application,Lcom/tencent/mm/plugin/appbrand/j...           u   
1026  <Application,Lcom/tencent/mm/plugin/appbrand/j...           u   

                                            Super Class  Number of Args  \
0                                                   NaN               0   
1                                                   NaN               0   
2     <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   
3                                                   NaN               0   
4     <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   
...                                                 ...             ...   
1022  <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   
1023  <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   
1024  <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   
1025  <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   
1026  <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   

                                         Original Class  
0                                 <Application,Lgt0/s0>  
1     <Application,Lcom/tencent/mm/plugin/appbrand/m...  
2                                 <Application,Lzt0/s1>  
3     <Application,Lcom/tencent/mm/plugin/appbrand/j...  
4     <Application,Lcom/tencent/mm/plugin/appbrand/j...  
...                                                 ...  
1022  <Application,Lcom/tencent/mm/plugin/appbrand/j...  
1023  <Application,Lcom/tencent/mm/plugin/appbrand/j...  
1024  <Application,Lcom/tencent/mm/plugin/appbrand/j...  
1025  <Application,Lcom/tencent/mm/plugin/appbrand/j...  
1026  <Application,Lcom/tencent/mm/plugin/appbrand/j...  

[1027 rows x 7 columns]

drop method not found and method name != u


In [13]:
# drop null values row where MethodName is null 
df = df.dropna(subset=['Method Name'])
df = df[df['Method Name'] == 'u']
df


Name     Return Type  \
2           showNavigationBarLoading  <Primordial,V>   
4                           saveFile  <Primordial,V>   
5                        openSetting  <Primordial,V>   
6               removeXWebLivePlayer  <Primordial,V>   
7                     searchContacts  <Primordial,V>   
...                              ...             ...   
1022         openChannelsOrderCenter  <Primordial,V>   
1023         openWCPayLQTDepositPlan  <Primordial,V>   
1024    startFacialRecognitionVerify  <Primordial,V>   
1025                   shareTimeline  <Primordial,V>   
1026  enableLocationUpdateBackground  <Primordial,V>   

                                             Class Name Method Name  \
2                                 <Application,Lzt0/s1>           u   
4     <Application,Lcom/tencent/mm/plugin/appbrand/j...           u   
5     <Application,Lcom/tencent/mm/plugin/appbrand/j...           u   
6     <Application,Lcom/tencent/mm/plugin/appbrand/e...           u   
7     <Application,Lcom/tencent/mm/plugin/appbrand/j...           u   
...                                                 ...         ...   
1022  <Application,Lcom/tencent/mm/plugin/appbrand/j...           u   
1023  <Application,Lcom/tencent/mm/plugin/appbrand/j...           u   
1024  <Application,Lcom/tencent/mm/plugin/appbrand/j...           u   
1025  <Application,Lcom/tencent/mm/plugin/appbrand/j...           u   
1026  <Application,Lcom/tencent/mm/plugin/appbrand/j...           u   

                                            Super Class  Number of Args  \
2     <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   
4     <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   
5     <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   
6     <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   
7     <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   
...                                                 ...             ...   
1022  <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   
1023  <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   
1024  <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   
1025  <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   
1026  <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   

                                         Original Class  
2                                 <Application,Lzt0/s1>  
4     <Application,Lcom/tencent/mm/plugin/appbrand/j...  
5     <Application,Lcom/tencent/mm/plugin/appbrand/j...  
6                                  <Application,Ldw0/h>  
7     <Application,Lcom/tencent/mm/plugin/appbrand/j...  
...                                                 ...  
1022  <Application,Lcom/tencent/mm/plugin/appbrand/j...  
1023  <Application,Lcom/tencent/mm/plugin/appbrand/j...  
1024  <Application,Lcom/tencent/mm/plugin/appbrand/j...  
1025  <Application,Lcom/tencent/mm/plugin/appbrand/j...  
1026  <Application,Lcom/tencent/mm/plugin/appbrand/j...  

[707 rows x 7 columns]

Superclass should be <Application,Lcom/tencent/mm/plugin/appbrand/jsapi/f>

In [14]:
df = df[df['Super Class'] == '<Application,Lcom/tencent/mm/plugin/appbrand/jsapi/f>']
df

Name     Return Type  \
2           showNavigationBarLoading  <Primordial,V>   
4                           saveFile  <Primordial,V>   
5                        openSetting  <Primordial,V>   
6               removeXWebLivePlayer  <Primordial,V>   
7                     searchContacts  <Primordial,V>   
...                              ...             ...   
1022         openChannelsOrderCenter  <Primordial,V>   
1023         openWCPayLQTDepositPlan  <Primordial,V>   
1024    startFacialRecognitionVerify  <Primordial,V>   
1025                   shareTimeline  <Primordial,V>   
1026  enableLocationUpdateBackground  <Primordial,V>   

                                             Class Name Method Name  \
2                                 <Application,Lzt0/s1>           u   
4     <Application,Lcom/tencent/mm/plugin/appbrand/j...           u   
5     <Application,Lcom/tencent/mm/plugin/appbrand/j...           u   
6     <Application,Lcom/tencent/mm/plugin/appbrand/e...           u   
7     <Application,Lcom/tencent/mm/plugin/appbrand/j...           u   
...                                                 ...         ...   
1022  <Application,Lcom/tencent/mm/plugin/appbrand/j...           u   
1023  <Application,Lcom/tencent/mm/plugin/appbrand/j...           u   
1024  <Application,Lcom/tencent/mm/plugin/appbrand/j...           u   
1025  <Application,Lcom/tencent/mm/plugin/appbrand/j...           u   
1026  <Application,Lcom/tencent/mm/plugin/appbrand/j...           u   

                                            Super Class  Number of Args  \
2     <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   
4     <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   
5     <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   
6     <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   
7     <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   
...                                                 ...             ...   
1022  <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   
1023  <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   
1024  <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   
1025  <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   
1026  <Application,Lcom/tencent/mm/plugin/appbrand/j...               4   

                                         Original Class  
2                                 <Application,Lzt0/s1>  
4     <Application,Lcom/tencent/mm/plugin/appbrand/j...  
5     <Application,Lcom/tencent/mm/plugin/appbrand/j...  
6                                  <Application,Ldw0/h>  
7     <Application,Lcom/tencent/mm/plugin/appbrand/j...  
...                                                 ...  
1022  <Application,Lcom/tencent/mm/plugin/appbrand/j...  
1023  <Application,Lcom/tencent/mm/plugin/appbrand/j...  
1024  <Application,Lcom/tencent/mm/plugin/appbrand/j...  
1025  <Application,Lcom/tencent/mm/plugin/appbrand/j...  
1026  <Application,Lcom/tencent/mm/plugin/appbrand/j...  

[707 rows x 7 columns]

In [28]:
names_all = df["Name"]

# convert the series into an array
names_all = names_all.values
# names_all = set(names_all)


names_all = set(names_all)
len((names_all))
# print duplciate names in names_all
# import collections
# print([item for item, count in collections.Counter(names_all).items() if count > 1])




960

Documented APIs

In [18]:
f = open("docAPIs.txt", "r")
names_doc = f.read().splitlines()
f.close()
# len(names_doc)
# names_doc = set(names_doc)
len(names_doc)

1280

In [29]:


# print all names that are present in names_all but not present in names_doc
undocumented = []
for name in names_all:
    if name not in names_doc:
        undocumented.append(name)

len(set(undocumented))

727

In [35]:

# write the undocumented APIs to a file
f = open("undocumentedAPIsStrict.txt", "w")
for name in undocumented:
    f.write(name + "\n")
f.close()


In [2]:
df = pd.read_csv('WechatApisParaupdated.csv')
df

Name   CtrlIndex  \
0                    requestPayment        0x39   
1                      getAppConfig        0x65   
2                          openCTID       0x4b9   
3                     shareGameLive       0x409   
4    shareFinderEndorsementToFriend  0xfffffffe   
..                              ...         ...   
702                       addMapArc  0xfffffffe   
703               requestSnsPayment       0x43f   
704           showImageOperateSheet       0x272   
705                    showKeyboard         0x1   
706            startBeaconDiscovery        0xdd   

                                                Method  \
0    < Application, Lcom/tencent/mm/plugin/appbrand...   
1    < Application, Lcom/tencent/mm/plugin/appbrand...   
2    < Application, Lcom/tencent/mm/plugin/appbrand...   
3    < Application, Lat0/k, u(Lcom/tencent/mm/plugi...   
4    < Application, Lcom/tencent/mm/plugin/appbrand...   
..                                                 ...   
702  < Application, Lit0/b, u(Lcom/tencent/mm/plugi...   
703  < Application, Lcom/tencent/mm/plugin/appbrand...   
704  < Application, Lcom/tencent/mm/plugin/appbrand...   
705  < Application, Lcom/tencent/luggage/game/jsapi...   
706  < Application, Ldt0/g, u(Lcom/tencent/mm/plugi...   

                                                 Class  \
0    <Application,Lcom/tencent/mm/plugin/appbrand/j...   
1    <Application,Lcom/tencent/mm/plugin/appbrand/j...   
2    <Application,Lcom/tencent/mm/plugin/appbrand/j...   
3                                 <Application,Lat0/k>   
4    <Application,Lcom/tencent/mm/plugin/appbrand/j...   
..                                                 ...   
702                               <Application,Lit0/c>   
703  <Application,Lcom/tencent/mm/plugin/appbrand/j...   
704  <Application,Lcom/tencent/mm/plugin/appbrand/j...   
705  <Application,Lcom/tencent/luggage/game/jsapi/k...   
706                               <Application,Ldt0/g>   

                                  Superclass  \
0    Lcom/tencent/mm/plugin/appbrand/jsapi/f   
1    Lcom/tencent/mm/plugin/appbrand/jsapi/f   
2    Lcom/tencent/mm/plugin/appbrand/jsapi/f   
3    Lcom/tencent/mm/plugin/appbrand/jsapi/f   
4    Lcom/tencent/mm/plugin/appbrand/jsapi/f   
..                                       ...   
702  Lcom/tencent/mm/plugin/appbrand/jsapi/f   
703  Lcom/tencent/mm/plugin/appbrand/jsapi/f   
704  Lcom/tencent/mm/plugin/appbrand/jsapi/f   
705  Lcom/tencent/mm/plugin/appbrand/jsapi/f   
706  Lcom/tencent/mm/plugin/appbrand/jsapi/f   

                                            Parameters  
0                                                   {}  
1                                       {"type":"Int"}  
2                           {"miniProgramID":"String"}  
3    {"Val":"String","ExptCheckSum":"String","EndTi...  
4    {"relativePath":"String","appid":"String","tit...  
..                                                 ...  
702  {"color":"String","pass":"JSONObject","latitud...  
703  {"Val":"String","package":"String","EndTime":"...  
704                                                 {}  
705  {"showStoreEmoticon":"Boolean","showEmoticonFi...  
706                                 {"uuids":"String"}  

[707 rows x 6 columns]

707 total APIs,(639 unique) with invariants method signature, superclass and superpackage , documented APIs = 1280

(check why are we getting duplicates)

In [13]:
names_all = df["Name"]
names_all = names_all.values
len(set(names_all))
# print([item for item, count in collections.Counter(names_all).items() if count > 1])


639

In [15]:
undocumented = []
for name in names_all:
    if name not in names_doc:
        undocumented.append(name)

len(undocumented)

500

In [44]:
# load a json file 
import json
f = open("undocStrict(manual).json", "r")
data = json.load(f)
data

{'updateXWebLivePusher': {},
 'enterWCCoinHomePage': {},
 'openWCPayOverseaPaymentReceive': {'errno': 102,
  'errMsg': 'openWCPayOverseaPaymentReceive:fail:access denied'},
 'insertNativeView': {},
 'launchApplicationDirectly': {'errno': 102,
  'errMsg': 'launchApplicationDirectly:fail:access denied'},
 'sendBizRedPacket': {'errno': 102,
  'errMsg': 'sendBizRedPacket:fail:access denied'},
 'updateNativeView': {},
 'splashAdAllViewLayoutStateChange': {},
 'NFCGetHistoricalBytes': {},
 'createFinderAccount': {},
 'getLocalLiveInfo': {},
 'setNavigateBackInterceptionForFakeNative': {},
 'extTransfer': {},
 'isSupportNFC': {},
 'updateXWebLivePlayer': {},
 'getAdPushMsg': {},
 'openChannelsPostPage': {},
 'shareGameLive': {},
 'addMapVisualLayer': {},
 'rtosWatchQuickReply': {},
 'insertMagicBrushFrame': {},
 'shareAppMessageToSpecificContactForFakeNativeDirectly': {},
 'openWCPayLQTDepositPlan': {},
 'getLastLocationCache': {},
 'deleteUserAutoFillData': {},
 'exitGameLive': {},
 'stopNFC

In [46]:
## Initialize three dictionaries
empty_api_dict = {}
access_denied_api_dict = {}
other_api_dict = {}

# Iterate through the JSON data
for api_name, api_data in data.items():
    if not api_data:
        empty_api_dict[api_name] = api_data
    elif "fail:access denied" in api_data.get('errMsg', ''):
        access_denied_api_dict[api_name] = api_data
    else:
        other_api_dict[api_name] = api_data

# Output the dictionaries
# print("Empty API Dict:", len(empty_api_dict))
# print("Access Denied API Dict:", len(access_denied_api_dict))
# print("Other API Dict:", other_api_dict)

empty_api_dict

{'updateXWebLivePusher': {},
 'enterWCCoinHomePage': {},
 'insertNativeView': {},
 'updateNativeView': {},
 'splashAdAllViewLayoutStateChange': {},
 'NFCGetHistoricalBytes': {},
 'createFinderAccount': {},
 'getLocalLiveInfo': {},
 'setNavigateBackInterceptionForFakeNative': {},
 'extTransfer': {},
 'isSupportNFC': {},
 'updateXWebLivePlayer': {},
 'getAdPushMsg': {},
 'openChannelsPostPage': {},
 'shareGameLive': {},
 'addMapVisualLayer': {},
 'rtosWatchQuickReply': {},
 'insertMagicBrushFrame': {},
 'shareAppMessageToSpecificContactForFakeNativeDirectly': {},
 'openWCPayLQTDepositPlan': {},
 'getLastLocationCache': {},
 'deleteUserAutoFillData': {},
 'exitGameLive': {},
 'stopNFCDiscovery': {},
 'cancelDownloadTaskForNative': {},
 'jumpToWCPayMessage': {},
 'operateDownloadTask': {},
 'stopBLEPeripheralAdvertising': {},
 'updateMapMarkers': {},
 'shareSpecificWxapp': {},
 'predownloadMiniProgramPackage': {},
 'requestSRFile': {},
 'editTextWithPopForm': {},
 'insertFinderWidget': {},

In [41]:
import json
f = open("undocStrict(manual).json", "r")
data = json.load(f)
data

{'updateXWebLivePusher': {},
 'enterWCCoinHomePage': {},
 'openWCPayOverseaPaymentReceive': {'errno': 102,
  'errMsg': 'openWCPayOverseaPaymentReceive:fail:access denied'},
 'insertNativeView': {},
 'launchApplicationDirectly': {'errno': 102,
  'errMsg': 'launchApplicationDirectly:fail:access denied'},
 'sendBizRedPacket': {'errno': 102,
  'errMsg': 'sendBizRedPacket:fail:access denied'},
 'updateNativeView': {},
 'splashAdAllViewLayoutStateChange': {},
 'NFCGetHistoricalBytes': {},
 'createFinderAccount': {},
 'getLocalLiveInfo': {},
 'setNavigateBackInterceptionForFakeNative': {},
 'extTransfer': {},
 'isSupportNFC': {},
 'updateXWebLivePlayer': {},
 'getAdPushMsg': {},
 'openChannelsPostPage': {},
 'shareGameLive': {},
 'addMapVisualLayer': {},
 'rtosWatchQuickReply': {},
 'insertMagicBrushFrame': {},
 'shareAppMessageToSpecificContactForFakeNativeDirectly': {},
 'openWCPayLQTDepositPlan': {},
 'getLastLocationCache': {},
 'deleteUserAutoFillData': {},
 'exitGameLive': {},
 'stopNFC